nn/module.go

```diff
@@ -1,38 +1,174 @@
 package nn
 
 import (
+	"fmt"
 	"log"
 	"reflect"
 
 	torch "github.com/wangkuiyi/gotorch"
 )
 
-// Module interface
-type Module interface {
-	Forward(x torch.Tensor) torch.Tensor
+// Dtype is the data type of scalars
+type Dtype int // TODO(shendiaomo): this is a placeholder to be defined later
+
+// DeviceType is the type of devices
+type DeviceType int // TODO(shendiaomo): this is a placeholder to be defined latert
+
+// DeviceIndex is the index of available devices
+type DeviceIndex int16
+
+// Device represents a a compute device on which a tensor is located.
+type Device struct {
+	typ DeviceType
+	idx DeviceIndex
+}
+
+// IModule is the interface of `Module`s
+type IModule interface {
+	// Train enables "training" mode
+	Train(on bool)
+	// IsTraining returns true if the module is in training mode
+	IsTraining() bool
+	// To recursively casts all parameters to the given `dtype` and `device`.
+	To(device Device, dtype Dtype, nonBlocking bool)
+	// ZeroGrad recursively zeros out the `grad` value of each registered parameter.
+	ZeroGrad()
+	// String is for printing modules prettily
+	String() string
+}
+
+// Module contains default implementation of `Module`s
+type Module struct {
+	outer IModule
+	// Whether the module is in training mode.
+	isTraining bool
+	// The module's name (e.g. "LSTM").
+	name string
+}
+
+// Init initializes a `Module`, using a `Module` that's not `Init`ed is undefined behavior
```

Comment:
We need to be able to check if a module is not initialized, and warns users about it. We cannot accept "undefined behavior".